main/src/ca/uwaterloo/flix/language/phase/HtmlDocumentor.scala

```diff
@@ -0,0 +1,640 @@
+
+    sb.append("<div class='doc'>")
+    sb.append(parsed)
+    sb.append("</div>")
+  }
+
+  /**
+    * Document the the given `Type`, `tpe`.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docType(tpe: Type)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append("<span class='type'>")
+    sb.append(esc(FormatType.formatType(tpe)))
+    sb.append("</span>")
+  }
+
+  /**
+    * Document the the given `Type`, `eff`, when it is known to be in effect position.
+    *
+    * The result will be appended to the given `StringBuilder`, `sb`.
+    */
+  private def docEffectType(eff: Type)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    sb.append("<span class='effect'>")
+    sb.append(esc(FormatType.formatType(eff)))
+    sb.append("</span>")
+  }
+
+  /**
+    * Runs the given `docElt` on each element of `list`, separated by the string: ", " (comma + space)
+    */
+  private def docList[T](list: List[T])(docElt: T => Unit)(implicit flix: Flix, sb: StringBuilder): Unit = {
+    for ((e, i) <- list.zipWithIndex) {
+      docElt(e)
+      if (i < list.length - 1) {
+        sb.append(", ")
+      }
+    }
+  }
+
+  private def clearOutputDirectory(): Unit = {
+    if (Files.exists(OutputDirectory)) {
+      Files.list(OutputDirectory).forEach {
+        f => Files.delete(f)
+      }
+    }
+  }
+
+  /**
+    * Make a copy of the stylesheet into the output directory.
+    */
+  private def writeStyles(): Unit = {
+    val source = Source.fromURL(getClass.getResource(Stylesheet))
+    writeFile("styles.css", source.mkString)
+  }
+
+  /**
+    * Write the documentation output string of the `Module`, `mod`, into the output directory with a suitable name.
+    */
+  private def writeModule(mod: Module, output: String): Unit = {
+    val name = if (mod.namespace.isEmpty) List(RootNS) else mod.namespace
+    writeFile(s"${name.mkString(".")}.html", output)
+  }
+
+  /**
+    * Write the file to the output directory with the given file name.
+    */
+  private def writeFile(name: String, output: String): Unit = {
+    val path = OutputDirectory.resolve(name)
+    try {
+      Files.createDirectories(OutputDirectory)
+      Files.writeString(path, output)
+    } catch {
+      case ex: IOException => throw new RuntimeException(s"Unable to write to path '$path'.", ex)
+    }
+  }
+
+  /**
+    * Create a raw link to the given `SourceLocation`.
+    */
+  private def createLink(loc: SourceLocation): String = {
+    // TODO make it also work for local user code
+    s"https://github.com/flix/flix/blob/master/main/src/library/${loc.source.name}#L${loc.beginLine}-L${loc.beginLine}"
+  }
+
+  /**
+    * Escape any HTML in the string.
+    */
+  private def esc(s: String): String = xml.Utility.escape(s)
+
+  /**
+    * A represention of a module that's easier to work with while generating documention.
+    */
+  private case class Module(namespace: List[String],
```

Comment:
I am a bit worried about this approach and below with class.

Because this means there is now duplicate information and more importantly, we have to maintain these data structures.

Why do you need them?

Reply:
It's just nicer to have all the relevant data in one structure mirroring the way that you would think about it, instead of having each `docX` function contain both the logic for extracting it from the `root` and generating the code.

Reply:
Okay. We can start this way and see what happens.